File: data structures and algorithms/python/08_020.py
import heapq

def process_evidence(evidence, hours_per_day):
   processing_order = []
   current_day = 0
   hours_left_today = hours_per_day
   
   # Priority queue as min heap with negative scores
   evidence_queue = []
   
   # Initial queue population
   for item in evidence:
      priority = (-(item["importance"] / 
               item["processing_time"]) * 
               (1 / item["days_until_expiration"]))
      heapq.heappush(evidence_queue, 
               (priority, item["id"], item))
   
   while evidence_queue:
      # Get highest priority evidence
      _, item_id, item = heapq.heappop(evidence_queue)
      
      # Check if it's expired
      if item["days_until_expiration"] <= current_day:
         continue
      
      # Process the evidence
      processing_order.append(item_id)
      
      # Update time tracking
      remaining_time = item["processing_time"]
      while remaining_time > 0:
         if hours_left_today >= remaining_time:
            hours_left_today -= remaining_time
            remaining_time = 0
         else:
            remaining_time -= hours_left_today
            current_day += 1
            hours_left_today = hours_per_day
            
            # Recalculate priorities for remaining evidence
            updated_queue = []
            while evidence_queue:
               _, id, ev = heapq.heappop(evidence_queue)
               if ev["days_until_expiration"] > current_day:
                  new_priority = (-(ev["importance"] / 
                           ev["processing_time"]) * 
                           (1 / (ev["days_until_expiration"] - 
                           current_day)))
                  heapq.heappush(updated_queue, 
                           (new_priority, id, ev))
            evidence_queue = updated_queue
   
   return processing_order

File: data structures and algorithms/cs/08_020.cs
using System;
using System.Collections.Generic;

class Evidence {
    public string Id { get; set; }
    public int Importance { get; set; }
    public int ProcessingTime { get; set; }
    public int DaysUntilExpiration { get; set; }

    public Evidence(string id, int importance, int processingTime, int daysUntilExpiration) {
        Id = id;
        Importance = importance;
        ProcessingTime = processingTime;
        DaysUntilExpiration = daysUntilExpiration;
    }

    public double GetPriority(int currentDay) {
        int daysLeft = DaysUntilExpiration - currentDay;
        if (daysLeft <= 0) return double.NegativeInfinity;
        return (Importance / (double)ProcessingTime) * (1.0 / daysLeft);
    }
}

public class EvidenceProcessor {
    public static List<string> ProcessEvidence(List<Evidence> evidence, int hoursPerDay) {
        var processingOrder = new List<string>();
        int currentDay = 0;
        int hoursLeftToday = hoursPerDay;

        var evidenceQueue = new PriorityQueue<Evidence>((e1, e2) => e2.GetPriority(currentDay).CompareTo(e1.GetPriority(currentDay)));
        evidenceQueue.AddRange(evidence);

        while (evidenceQueue.Count > 0) {
            var item = evidenceQueue.Dequeue();

            if (item.DaysUntilExpiration <= currentDay) {
                continue;
            }

            processingOrder.Add(item.Id);

            int remainingTime = item.ProcessingTime;
            while (remainingTime > 0) {
                if (hoursLeftToday >= remainingTime) {
                    hoursLeftToday -= remainingTime;
                    remainingTime = 0;
                } else {
                    remainingTime -= hoursLeftToday;
                    currentDay++;
                    hoursLeftToday = hoursPerDay;

                    var remaining = new List<Evidence>();
                    while (evidenceQueue.Count > 0) {
                        remaining.Add(evidenceQueue.Dequeue());
                    }

                    evidenceQueue = new PriorityQueue<Evidence>((e1, e2) => e2.GetPriority(currentDay).CompareTo(e1.GetPriority(currentDay)));
                    foreach (var e in remaining) {
                        if (e.DaysUntilExpiration > currentDay) {
                            evidenceQueue.Enqueue(e);
                        }
                    }
                }
            }
        }

        return processingOrder;
    }
}